File: app/src/main/java/com/ewaldmire/osmride/weight/NavyBodyFat.kt
package com.ewaldmire.osmride.weight

import kotlin.math.log10

/** U.S. Navy circumference method body-fat % estimate. Male formula only - this app is scoped to
 * a single user, per the body metrics feature's own spec, so there's no gender selector to build. */
object NavyBodyFat {
    /** Returns null for physically nonsensical inputs (e.g. neck >= waist) rather than a garbage
     * number - log10 of a non-positive value is undefined. */
    fun estimatePercent(waistCm: Double, neckCm: Double, heightCm: Double): Double? {
        val waistMinusNeck = waistCm - neckCm
        if (waistMinusNeck <= 0 || heightCm <= 0) return null
        val denominator = 1.0324 - 0.19077 * log10(waistMinusNeck) + 0.15456 * log10(heightCm)
        if (denominator <= 0) return null
        return 495.0 / denominator - 450.0
    }
}

File: linux-client/osm_ride_linux/weight/navy_body_fat.py
"""Mirrors app/src/main/java/com/ewaldmire/osmride/weight/NavyBodyFat.kt."""

from __future__ import annotations

import math


def estimate_percent(waist_cm: float, neck_cm: float, height_cm: float) -> float | None:
    """Navy method body-fat % estimate (metric, male-only - this app is scoped to a single user).
    Returns None for physically nonsensical inputs (waist <= neck, non-positive denominator)."""
    waist_minus_neck = waist_cm - neck_cm
    if waist_minus_neck <= 0 or height_cm <= 0:
        return None
    denominator = 1.0324 - 0.19077 * math.log10(waist_minus_neck) + 0.15456 * math.log10(height_cm)
    if denominator <= 0:
        return None
    return 495.0 / denominator - 450.0
